src/include/hardware.h

```diff
@@ -45,6 +45,29 @@
 #define WT_READ_ONCE(v, val) WT_ORDERED_READ(v, val)
 #endif
 
+/*
+ * WT_WRITE_ONCE --
+ *
+ * Ensure a single write to memory in the source code produces a single write to memory in the
+ * compiled output.
+ *
+ * We can describe a scenario where a variable in memory is read to a local register, and then the
+ * register value is repeatedly updated before the final value is written back to memory. Without a
+ * WT_WRITE_ONCE wrapping the final write back to memory, the compiler is allowed to convert these
+ * register writes into writes to memory (see "register spilling") which can leak interim, invalid,
+ * state to other threads. We call these 'invented stores' in WiredTiger.
```

Comment:
WiredTiger generally reads into a local variable, and changes the local variable to protect against this. Then finally assigns back to the shared variable in a single assignment.

Is this proposing that the approach isn't adequate, that a volatile cast is necessary to ensure intermediate writes to a local variable don't make it back to a shared variable?

I'd also like to understand how this overlaps with our 64-bit write requirement (https://github.com/wiredtiger/wiredtiger/blame/develop/src/include/gcc.h#L53).

Reply:
I don't have a hard example of this occurring, but [here](https://godbolt.org/z/YeKeTTEqb)'s a naive proof of concept where the stack variable is dropped in favour of using a register. Combining this with stack spilling we _should_ spill to the stack variables memory location, but I don't know how to guarantee that. Using `WRITE_ONCE` closes the door on that conversation and guarantees a single write to shared memory.

> I'd also like to understand how this overlaps with our 64-bit write requirement

I don't think it does. The 64-bit write requirement avoids store tearing while WRITE_ONCE is preventing write inventions.

Reply:
Andrew is correct about the intention of WRITE_ONCE here.